documentation/src/main/asciidoc/reference/search-dsl-projection.asciidoc

```diff
@@ -740,6 +740,86 @@ include::{sourcedir}/org/hibernate/search/documentation/search/projection/Projec
 ----
 ====
 
+[[search-dsl-projection-highlight]]
+== `highlight`: return highlighted field values from matched documents
+
+The `highlight` projection returns fragments from full-text fields of matched documents that caused a query match.
+
+[IMPORTANT]
+====
+In order to use highlight projections of a given field,
+you need to provide the list of <<mapping-directfieldmapping-highlightable,highlighters supported by the field>> in the mapping.
+
+The <<mapping-directfieldmapping-highlightable,highlightable>> default already can enable the support of highlighting
+in some cases. For more details see <<mapping-directfieldmapping-highlightable-default, how the `DEFAULT` highlightable value behaves>>.
+====
+
+
+[[search-dsl-projection-highlight-syntax]]
+=== Syntax
+
+The `highlight` projection always returns a list of string values per highlighted field, no matter if the field is
+a single or a multivalued one:
+
+.Returning highlights for matched documents
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/search/projection/ProjectionDslIT.java[tags=highlight]
+----
+====
+
+[[search-dsl-projection-highlight-multivalued]]
+=== Multivalued fields
+
+Each value of a multivalued field is being highlighted. See <<search-dsl-highlighting-highlighter-options, how highlighter can be configured>>
+to adjust the returned results' behaviour and structure.
+
+[IMPORTANT]
+====
+At this moment, highlights on fields of nested objects is not supported, however, they can work on fields of nested objects
+with <<mapping-indexedembedded-structure-flattened,flattened structure>>.
```

Comment:
mm yeah .. so if we have `ObjectStructure.NESTED`, things go badly for the highlighting... the query we get is some version of  `ToParentBlockJoinQuery`, and highlighters have no idea what to do about it (a more detailed example from our chat - https://hibernate.zulipchat.com/#narrow/stream/132092-hibernate-search-dev/topic/HSEARCH-2192.20Highlighting/near/345345408) which results in an empty highlight list. Same thing for ES - just an empty list. So it really works fine only with `ObjectStructure.FLATTENED`.

Reply:
Ah, that's... underwhelming. Do we have an issue (improvement) about this? Because we need one.

Anyway, in that case:

```suggestion
At the moment, highlighting a field within a <<mapping-indexedembedded-structure-nested,nested object>> is link:TODO_LINK_TO_JIRA_ISSUE[not supported] and attempting to do so will lead to an exception.

 Highlighting a field within a <<mapping-indexedembedded-structure-flattened,flattened object>> will work correctly.
```

Reply:
This behavior is not really what I had in mind though, I was more expecting nesting to be ignored:

> > Alright, would it be simpler to ignore nesting in both backends and just act like you currently do on Elasticsearch?
> 
> Basically we'd forbid using .highlight in a nested context (see [HSEARCH-4827](https://hibernate.atlassian.net/browse/HSEARCH-4827)), and highlighting would always apply to all nested objects

If even that doesn't work, you should probably enforce stricter checks. Like throwing an exception in `org.hibernate.search.backend.lucene.search.projection.impl.LuceneFieldHighlightProjection.Factory#create` if  `field.nestedDocumentPath() != null` (and same for Elasticsearch).

Reply:
ok, will do! 😃 
To support either this case or the `object()` one, we'd still most likely need to work with Elasticsearch's `inner_hits.highlight` . And with Lucene.... it is the problem with queries in both scenarios, so I assume if we get one to work - the other will work too. Here's the jira - HSEARCH-4841